common/client-libs/gateway-client/src/client.rs

```diff
@@ -379,21 +379,26 @@ impl GatewayClient {
                 ws_stream,
                 self.local_identity.as_ref(),
                 self.gateway_identity,
-                self.coconut_credential.clone(),
             )
             .await
             .map_err(GatewayClientError::RegistrationFailure),
             _ => unreachable!(),
         }?;
-
-        self.authenticated = true;
-        Ok(shared_key)
+        self.authenticated = match self.read_control_response().await? {
+            ServerResponse::Register { status } => Ok(status),
+            ServerResponse::Error { message } => Err(GatewayClientError::GatewayError(message)),
+            _ => Err(GatewayClientError::NotAuthenticated),
+        }?;
+        if self.authenticated {
+            self.shared_key = Some(Arc::new(shared_key));
```

Comment:
Because the returned `status` of the `ServerResponse::Register` might be `false` for whatever reason, in which case the state is not authenticated and the shared key is not valid. This clearly doesn't happen right now, as we do all of the needed checks in the handshake communication, but in terms of the returned `ServerResponse::Register { status }`, it might be possible to fail _after_ all the handshake work.